p2pserver/link/link.go

```diff
@@ -190,20 +258,66 @@ func (this *Link) needSendMsg(msg types.Message) bool {
 }
 
 //addReqRecord add request record by removing outdated request records
-func (this *Link) addReqRecord(msg types.Message) {
+func (self *Link) addReqRecord(msg types.Message) {
 	if msg.CmdType() != common.GET_DATA_TYPE {
 		return
 	}
 	now := time.Now().Unix()
-	if len(this.reqRecord) >= common.MAX_REQ_RECORD_SIZE-1 {
-		for id := range this.reqRecord {
-			t := this.reqRecord[id]
+	if len(self.reqRecord) >= common.MAX_REQ_RECORD_SIZE-1 {
+		for id := range self.reqRecord {
+			t := self.reqRecord[id]
 			if int(now-t) > common.REQ_INTERVAL {
-				delete(this.reqRecord, id)
+				delete(self.reqRecord, id)
 			}
 		}
 	}
 	var dataReq = msg.(*types.DataReq)
 	reqID := fmt.Sprintf("%x%s", dataReq.DataType, dataReq.Hash.ToHexString())
-	this.reqRecord[reqID] = now
+	self.reqRecord[reqID] = now
+}
+
+const sendBufSize = 64 * 1024
+
+func (self *Link) sendLoop() {
+	buffers := make([]byte, 0, sendBufSize)
+	var results []chan error
+	buffList := make([]buffData, 0, 64)
+	for {
+		owned, sealed := self.sendBuffer.buffers.Take()
+		if sealed {
+			return
+		}
+		buffList = getBufferData(buffList[:0], owned)
+		if len(buffList) > 0 {
+			for i := len(buffList) - 1; i >= 0; i -= 1 {
+				buffers = append(buffers, buffList[i].data...)
+				if buffList[i].result != nil {
+					results = append(results, buffList[i].result)
+				}
+				if len(buffers) >= sendBufSize/2 || i == 0 {
+					err := self.writeToConn(buffers)
+					self.sendBuffer.IncrBuffSize(-len(buffers))
+					for _, c := range results {
+						c <- err
+					}
+					if err != nil {
+						return
+					}
+					buffers = buffers[:0]
+					results = results[:0]
+				}
+			}
+		} else {
+			// no buffer has been taken, yield this goroutine to avoid busy loop
+			runtime.Gosched()
```

Comment:
can this do with channel?

Reply:
It will cause sending a notification for every msg to wake the writer, which will hurt performance in high contention scenario.